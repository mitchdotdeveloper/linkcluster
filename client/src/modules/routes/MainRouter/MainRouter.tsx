import React, { FC } from 'react';
import { Route, Switch } from 'react-router-dom';
import { AuthRouter } from '../AuthRouter/AuthRouter';
import { UserRouter } from '../UserRouter/UserRouter';
import { ProtectedRoute } from 'modules/helpers/ProtectedRoute/ProtectedRoute';

export type MainRouterProps = {};

export const MainRouter: FC<MainRouterProps> = () => {
  return (
    <Switch>
      <Route path="/auth">
        <AuthRouter />
      </Route>
      <ProtectedRoute path="/user">
        <UserRouter />
      </ProtectedRoute>
    </Switch>
  );
};
