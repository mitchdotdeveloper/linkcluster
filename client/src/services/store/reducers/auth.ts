import produce from 'immer';
import {
  DELETE_LOGOUT_FAILURE,
  DELETE_LOGOUT_REQUEST,
  DELETE_LOGOUT_SUCCESS,
  POST_LOGIN_FAILURE,
  POST_LOGIN_REQUEST,
  POST_LOGIN_SUCCESS,
  POST_REFRESH_FAILURE,
  POST_REFRESH_REQUEST,
  POST_REFRESH_SUCCESS,
  POST_SIGNUP_FAILURE,
  POST_SIGNUP_REQUEST,
  POST_SIGNUP_SUCCESS,
} from 'services/api/auth/actions';
import { AuthActionTypes, AuthStore } from 'types/services/store/auth';

const initialState: AuthStore = {
  jwt: '',
  isAuthenticated: false,
  error: null,
};

export const Auth = (state = initialState, action: AuthActionTypes) =>
  produce(state, (draft) => {
    switch (action.type) {
      case POST_SIGNUP_REQUEST:
      case POST_LOGIN_REQUEST:
      case DELETE_LOGOUT_REQUEST:
      case POST_REFRESH_REQUEST: {
        return state;
      }
      case POST_SIGNUP_SUCCESS:
      case POST_LOGIN_SUCCESS:
      case POST_REFRESH_SUCCESS: {
        const { jwt } = action.payload;

        draft.jwt = jwt;
        draft.isAuthenticated = true;
        draft.error = null;

        break;
      }
      case DELETE_LOGOUT_SUCCESS: {
        return initialState;
      }
      case POST_SIGNUP_FAILURE:
      case POST_LOGIN_FAILURE:
      case DELETE_LOGOUT_FAILURE:
      case POST_REFRESH_FAILURE: {
        const { error } = action.payload;

        draft.isAuthenticated = false;
        draft.error = error;

        break;
      }
    }
  });
