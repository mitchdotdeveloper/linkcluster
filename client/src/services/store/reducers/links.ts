import produce from 'immer';
import {
  GET_LINKS_FAILURE,
  GET_LINKS_REQUEST,
  GET_LINKS_SUCCESS,
} from 'services/api/links/actions';
import { LinksActionTypes, LinksStore } from 'types/services/store/links';

const initialState: LinksStore = {
  links: [],
  error: null,
};

export const Links = (state = initialState, action: LinksActionTypes) =>
  produce(state, (draft) => {
    switch (action.type) {
      case GET_LINKS_REQUEST: {
        return state;
      }
      case GET_LINKS_SUCCESS: {
        const { links } = action.payload;

        draft.links = links;
        draft.error = null;

        break;
      }
      case GET_LINKS_FAILURE: {
        const { error } = action.payload;

        draft.error = error;

        break;
      }
    }
  });
