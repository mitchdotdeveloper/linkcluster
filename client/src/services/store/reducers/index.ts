import { combineReducers } from 'redux';
import { User } from 'services/store/reducers/user';
import { Auth } from 'services/store/reducers/auth';
import { Links } from 'services/store/reducers/links';

export const rootReducer = combineReducers({ User, Auth, Links });

export type RootState = ReturnType<typeof rootReducer>;
