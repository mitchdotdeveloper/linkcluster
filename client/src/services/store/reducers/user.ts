import produce from 'immer';
import _merge from 'lodash.merge';
import {
  GET_USER_FAILURE,
  GET_USER_REQUEST,
  GET_USER_SUCCESS,
  UPDATE_USER_STORE_FROM_AUTH,
} from 'services/api/user/actions';
import { UserActionTypes, UserStore } from 'types/services/store/user';

const initialState: UserStore = {
  userID: null,
  user: null,
  error: null,
};

export const User = (state = initialState, action: UserActionTypes) =>
  produce(state, (draft) => {
    switch (action.type) {
      case GET_USER_REQUEST: {
        return state;
      }
      case GET_USER_SUCCESS: {
        const { user } = action.payload;

        draft.userID = user.userID;
        draft.user = user;
        draft.error = null;

        break;
      }
      case GET_USER_FAILURE: {
        const { error } = action.payload;

        draft.error = error;

        break;
      }
      case UPDATE_USER_STORE_FROM_AUTH: {
        const { user } = action.payload;

        if (user.userID) draft.userID = user.userID;
        draft.user = _merge(draft.user, user);

        break;
      }
    }
  });
