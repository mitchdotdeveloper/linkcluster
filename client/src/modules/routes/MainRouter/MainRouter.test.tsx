import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import { MainRouter } from './MainRouter';
import { RootState } from '../../../services/store/reducers/index';
import { Provider } from 'react-redux';

describe('MainRouter Suite', () => {
  const mockStore = configureMockStore<RootState>();

  test('Renders /auth route', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['/auth']}>
        <MainRouter />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();

    cleanup();
  });

  test('Renders /user route', () => {
    const store = mockStore({
      Auth: {
        isAuthenticated: true,
      },
      User: {
        user: { username: 'username' },
      },
    } as RootState);

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/user']}>
          <MainRouter />
        </MemoryRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();

    cleanup();
  });
});
