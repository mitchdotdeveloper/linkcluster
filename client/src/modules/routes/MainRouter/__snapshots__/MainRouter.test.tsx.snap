// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainRouter Suite Renders /auth route 1`] = `
<DocumentFragment>
  /auth/login
</DocumentFragment>
`;

exports[`MainRouter Suite Renders /user route 1`] = `
<DocumentFragment>
  username
</DocumentFragment>
`;
